File: api/models/booking.go
package models

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

type BookingStatus string

const (
	BookingPending   BookingStatus = "pending"
	BookingConfirmed BookingStatus = "confirmed"
	BookingCancelled BookingStatus = "cancelled"
)

type Booking struct {
	ID              uuid.UUID     `gorm:"type:char(36);primaryKey" json:"id"`
	UserID          uuid.UUID     `gorm:"type:char(36);not null" json:"user_id"`
	BoardingHouseID uuid.UUID     `gorm:"type:char(36);not null" json:"boarding_house_id"`
	StartDate       time.Time     `gorm:"type:date;not null" json:"start_date"`
	EndDate         time.Time     `gorm:"type:date;not null" json:"end_date"`
	Status          BookingStatus `gorm:"type:enum('pending','confirmed','cancelled');default:'pending'" json:"status"`

	User          *User          `gorm:"foreignKey:UserID" json:"user,omitempty"`
	BoardingHouse *BoardingHouse `gorm:"foreignKey:BoardingHouseID" json:"boarding_house,omitempty"`

	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"-"`
}

// Auto generate UUID before creation
func (b *Booking) BeforeCreate(tx *gorm.DB) (err error) {
	if b.ID == uuid.Nil {
		b.ID = uuid.New()
	}
	return
}

// BookingRequest represents the request payload for creating or updating a Booking
// Pastikan StartDate dan EndDate bertipe string agar bisa di-parse dari JSON
// Contoh: "start_date": "2024-07-01", "end_date": "2024-07-31"
type BookingRequest struct {
	BoardingHouseID string `json:"boarding_house_id" binding:"required"`
	StartDate       string `json:"start_date" binding:"required"`
	EndDate         string `json:"end_date" binding:"required"`
}

// BookingResponse represents the response payload for a Booking
type BookingResponse struct {
	ID              string        `json:"id"`
	UserID          string        `json:"user_id"`
	BoardingHouseID string        `json:"boarding_house_id"`
	StartDate       time.Time     `json:"start_date"`
	EndDate         time.Time     `json:"end_date"`
	Status          BookingStatus `json:"status"`
	CreatedAt       time.Time     `json:"created_at"`
	UpdatedAt       time.Time     `json:"updated_at"`

	User          *UserResponse          `json:"user,omitempty"`
	BoardingHouse *BoardingHouseResponse `json:"boarding_house,omitempty"`
}

File: mobile_application/lib/models/booking.dart
import 'package:mobile_application/models/boarding_house.dart';
import 'package:mobile_application/models/user.dart';

class Booking {
  final String id;
  final String userId;
  final String boardingHouseId;
  final String startDate;
  final String endDate;
  final String status;
  final String createdAt;
  final String updatedAt;
  final User user;
  final BoardingHouse boardingHouse;

  Booking({
    required this.id,
    required this.userId,
    required this.boardingHouseId,
    required this.startDate,
    required this.endDate,
    required this.status,
    required this.createdAt,
    required this.updatedAt,
    required this.user,
    required this.boardingHouse,
  });

  factory Booking.fromJson(Map<String, dynamic> json) {
    return Booking(
      id: json['id'] ?? '',
      userId: json['user_id'] ?? '',
      boardingHouseId: json['boarding_house_id'] ?? '',
      startDate: json['start_date'] ?? '',
      endDate: json['end_date'] ?? '',
      status: json['status'] ?? '',
      createdAt: json['created_at'] ?? '',
      updatedAt: json['updated_at'] ?? '',
      user: User.fromJson(json['user'] ?? {}),
      boardingHouse: BoardingHouse.fromJson(json['boarding_house'] ?? {}),
    );
  }
}

enum BookingStatus { pending, confirmed, cancelled }

BookingStatus bookingStatusFromString(String status) {
  switch (status) {
    case 'pending':
      return BookingStatus.pending;
    case 'confirmed':
      return BookingStatus.confirmed;
    case 'cancelled':
      return BookingStatus.cancelled;
    default:
      return BookingStatus.pending;
  }
}

String bookingStatusToString(BookingStatus status) {
  switch (status) {
    case BookingStatus.pending:
      return 'pending';
    case BookingStatus.confirmed:
      return 'confirmed';
    case BookingStatus.cancelled:
      return 'cancelled';
  }
}
